lib/ansible/plugins/action/gather_facts.py

```diff
@@ -0,0 +1,97 @@
+# Copyright (c) 2017 Ansible Project
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+import os
+import time
+
+from ansible.module_utils.common._collections_compat import MutableMapping
+
+from ansible import constants as C
+from ansible.plugins.action import ActionBase
+from ansible.utils.vars import combine_vars
+
+
+class ActionModule(ActionBase):
+
+    def _get_module_args(self, fact_module, task_vars):
+        mod_args = task_vars.get('ansible_facts_modules', {}).get(fact_module, {})
+        if isinstance(mod_args, MutableMapping):
+            mod_args.update(self._task.args.copy())
+        else:
+            mod_args = self._task.args.copy()
+
+        if fact_module != 'setup':
+            subset = mod_args.pop('gather_subset', None)
+            if subset not in ('all', ['all']):
+                self._display.warning('Ignoring subset(%s) for %s' % (subset, fact_module))
+
+        return mod_args
+
+    def run(self, tmp=None, task_vars=None):
+
+        self._supports_check_mode = True
+
+        result = super(ActionModule, self).run(tmp, task_vars)
+        result['ansible_facts'] = {}
+
+        force_serialization = task_vars.pop('ansible_facts_serialized', self._task.args.pop('force_serialization', None))
+
+        modules = task_vars.get('ansbile_facts_modules', {}).keys()
+        override_vars = {}
+        if modules:
+            override_vars['ansible_facts_modules'] = modules
+        modules = C.config.get_config_value('FACTS_MODULES', variables=override_vars)
+
+        jobs = {}
+        if force_serialization or len(modules) == 1:
+            # serially execute each module
+            for fact_module in modules:
+                # just one module, no need for fancy async
+                mod_args = self._get_module_args(fact_module, task_vars)
+                result.update(self._execute_module(module_name=fact_module, module_args=mod_args, task_vars=task_vars, wrap_async=False))
```

Comment:
The serial path doesn't look like it was updated to use combine_vars, so looks like you'd lose all but the last one here if > 1. Also, the async path only keeps the well-known status keys + `ansible_facts`, while the sync path keeps everything (previous error aside).

Reply:
Unless I'm missing something, this is still a problem... 2 modules, executed serially, each returning an ansible_facts sub-dict. Module 1's sub-dict will be replaced by module 2's, not merged, right? Which means you effectively only get facts from the last module to run... Right?

Reply:
no, i just have not finished updating the 'serial side'